ethexe/validator/src/lib.rs

```diff
@@ -16,99 +16,437 @@
+        } = request;
+
+        if db
+            .block_end_state_is_valid(block_hash)
+            .unwrap_or(false)
+            .not()
+        {
+            return Err(anyhow!(
+                "Requested block {block_hash} is not processed by this node"
+            ));
+        }
 
-        // broadcast (aggregated_code_commitments, aggregated_transitions_commitments) to the network peers
-        let commitments = (self.codes_aggregation()?, self.blocks_aggregation()?);
-        sender.publish_commitments((origin, commitments).encode());
+        if db
+            .block_outcome(block_hash)
+            .ok_or(anyhow!("Cannot get from db outcome for block {block_hash}"))?
+            .iter()
+            .map(ToDigest::to_digest)
+            .collect::<Digest>()
+            .ne(&transitions_digest)
+        {
+            return Err(anyhow!("Requested and local transitions digest mismatch"));
+        }
+
+        if db
+            .block_prev_commitment(block_hash)
+            .ok_or(anyhow!(
+                "Cannot get from db previous commitment block for block {block_hash}"
+            ))?
+            .ne(&allowed_prev_commitment_hash)
+        {
+            return Err(anyhow!(
+                "Requested and local previous commitment block hash mismatch"
+            ));
+        }
+
+        if Self::verify_is_predecessor(db, allowed_pred_block_hash, block_hash, None)?.not() {
+            return Err(anyhow!(
+                "{block_hash} is not a predecessor of {allowed_pred_block_hash}"
+            ));
+        }
 
         Ok(())
     }
 
-    pub fn clear(&mut self) {
-        self.current_codes.clear();
-        self.current_blocks.clear();
+    /// Verify whether `pred_hash` is a predecessor of `block_hash` in the chain.
+    fn verify_is_predecessor(
+        db: &impl BlockMetaStorage,
+        block_hash: H256,
+        pred_hash: H256,
+        max_distance: Option<u32>,
+    ) -> Result<bool> {
+        if block_hash == pred_hash {
+            return Ok(true);
+        }
+
+        let block_header = db
+            .block_header(block_hash)
+            .ok_or(anyhow!("header not found for block: {block_hash}"))?;
+
+        if block_header.parent_hash == pred_hash {
+            return Ok(true);
+        }
+
+        let pred_height = db
+            .block_header(pred_hash)
+            .ok_or(anyhow!("header not found for pred block: {pred_hash}"))?
+            .height;
+
+        let distance = block_header.height.saturating_sub(pred_height);
+        if max_distance.map(|d| d < distance).unwrap_or(false) {
+            return Err(anyhow!("distance is too large: {distance}"));
+        }
+
+        let mut block_hash = block_hash;
+        for _ in 0..=distance {
+            if block_hash == pred_hash {
+                return Ok(true);
+            }
+            block_hash = db
+                .block_header(block_hash)
+                .ok_or(anyhow!("header not found for block: {block_hash}"))?
```

Comment:
consider using ok_or_else in loop

Reply:
Ok, but I'm sure compiler would optimise this moment

Reply:
Done